### ISSUE TITLE
Unclear description of data protection key storage in Azure App Service

### ISSUE THREAD

**Issue by jennings** [dotnet/AspNetCore.Docs#28316]
I find the description of Data Protection key management for Azure App Service to be a little confusing. This part is hard to understand:

> When you swap between deployment slots, for example swapping Staging to Production or using A/B testing, any app using Data Protection won't be able to decrypt stored data using the key ring inside the previous slot.

It's not clear to me which behavior is being described:

1. Deployment keys swap between slots. If the software previously running in the Production slot encrypted data, the new software swapped into the Production slot will not be able to decrypt it.

2. Deployment keys are sticky to the slot. Your Production app will not be able to decrypt anything that it protected while it was running in the Staging slot.

I would find this clearer if it explicitly said either, "Data protection keys are swapped between slots," or "Data protection keys are sticky and remain in the slot they were created." I _think_ the actual behavior is that keys are swapped, because otherwise I don't see how users would be "logged out of an app that uses the standard ASP.NET Core cookie authentication," since production and staging slots use different URLs and therefore usually don't share cookies.

The next few sentences discuss "slot-independent key rings", but that sounds like it's describing sharing the same key ring with multiple slots. I do want separate key rings for each slot, I just would prefer them to be sticky.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 945c28b7-6cc7-d4ed-3b96-a401eef42f59
* Version Independent ID: eb45fbf9-29fb-c67b-2f50-454380a502dd
* Content: [Data Protection key management and lifetime in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/security/data-protection/configuration/default-settings?view=aspnetcore-7.0)
* Content Source: [aspnetcore/security/data-protection/configuration/default-settings.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/security/data-protection/configuration/default-settings.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

**Comment by adityamandaleeka** [comment]
cc: @bradygaster @amcasey PTAL

**Comment by Rick-Anderson** [comment]
> cc: @bradygaster @amcasey PTAL

email reminder sent.

**Comment by bradygaster** [comment]
See the email, hadn't seen this issue - thanks! Let me check on something via email first.

**Comment by bradygaster** [comment]
App Service has a special case where data protection stores keysin the app service local folder which gets synced between all instances. App service keeps this folder isolated by slots, so slot A has a different folder than slot B. When you swap slots the keys specific to that slot is used, so all cookies etc. get invalidated, and if you’ve manually protected data that data can no longer be decrypted. The app storage file system setup is well documented, [https://github.com/projectkudu/kudu/wiki/Understanding-the-Azure-App-Service-file-system](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fprojectkudu%2Fkudu%2Fwiki%2FUnderstanding-the-Azure-App-Service-file-system&data=05%7C01%7Cbradyg%40microsoft.com%7Cea6a5d77d5bc407a181108db1068572b%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638121812549696078%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=cVu4dTWAjSabMbQ9huWk8PjfkLGJ%2FhWMwzIn9c8sUIA%3D&reserved=0) and you can actually go see the code where we branch based on detecting we’re in azure. 

Would this level of explicitness be appropriate for the doc, you think? 

Also - if you *want* to facilitate a shared-key scenario, that's entirely possible using blob storage. That way all the machines would have their own key. We have docs on this already, but, I'd be keen to adding that sort of commentary to this doc if @Rick-Anderson thinks it'd be valuable to show customers how to do this on their own.

**Comment by jennings** [comment]
> App service keeps this folder isolated by slots, so slot A has a different folder than slot B.

This part is very clear in the documentation today.

> When you swap slots the keys specific to that slot is used, so all cookies etc. get invalidated, and if you’ve manually protected data that data can no longer be decrypted.

I still think this can be interpreted either way. What do you mean by, "the keys specific to that slot is used"? Are you saying that the keys are sticky to the slot (like sticky app settings), or do the keys swap slots just like the application code does?

I would find it less ambiguous if you describe _where the keys go_, not just what the _result to your application_ will be. One of these two things happen:

1. Data protection keys are sticky, so when you swap slots, the deployment keys are _not_ swapped. A data protection key generated in the production slot remain in the production slot even after swapping.
2. Data protection keys not sticky, get swapped when slots are swapped. After swapping, keys generated in the production slot swap to the staging slot and vice versa.

**Comment by jennings** [comment]
I think part of my confusion comes from the metaphor "sticky" not matching how the swap is actually implemented (maybe this is why the API calls it "sticky" but the Azure Portal calls it a "deployment slot setting").

Before the swap:

- Traffic for `production.example.com` is routed to Machine A (assuming a scale of 1 instance)
- Traffic for `staging.example.com` is routed to Machine B

Swapping really means changing the routing rules. After the swap:

- Traffic for `production.example.com` is routed to Machine B
- Traffic for `staging.example.com` is routed to Machine A

This means that so-called "sticky" settings are _actually_ moved from one machine to the other: Machine B is reconfigured with the sticky settings for production and Machine A is reconfigured with those for staging.

**Comment by bradygaster** [comment]
I agree that is a little confusing. "Sticky" to me seems to mean "the settings stick to the slot," not to the machine. Either way - this does make some sense to get some specific commentary in a doc about how the settings *and* the keys move. Adding @wadepickett and @alexwolfmsft to this discussion - I think we have an opportunity to make this a bit more explicit.
